<a href="https://colab.research.google.com/github/aabedaaiyyub/aabedaaiyyub.github.io/blob/main/f1Crypto.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Installing yfinance as it's more stable for Google Colab/Regional restrictions
!pip install yfinance pandas numpy matplotlib seaborn

In [ ]:
import yfinance as yf
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Tickers for the top 50 based on your list (standardizing for Yahoo Finance)
tickers = [
    'BTC-USD', 'ETH-USD', 'BNB-USD', 'XRP-USD', 'SOL-USD', 'TRX-USD',
    'DOGE-USD', 'ADA-USD', 'BCH-USD', 'XMR-USD', 'LINK-USD', 'XLM-USD',
    'LTC-USD', 'AVAX-USD', 'HBAR-USD', 'SHIB-USD', 'TON1-USD', 'CRO-USD',
    'DOT-USD', 'UNI7-USD', 'MNT-USD', 'BGB-USD', 'TAO-USD', 'AAVE-USD',
    'OKB-USD', 'NEAR-USD', 'ETC-USD', 'ONDO-USD', 'ASTR-USD'
]

# Fetching 1 year of daily data
print("Fetching data...")
data = yf.download(tickers, period="1y", interval="1d")['Close']

Fetching data...


/tmp/ipython-input-74572533.py:18: FutureWarning: YF.download() has changed argument auto_adjust default to True
  data = yf.download(tickers, period="1y", interval="1d")['Close']
[********              17%                       ]  5 of 29 completedERROR:yfinance:HTTP Error 404: {"quoteSummary":{"result":null,"error":{"code":"Not Found","description":"Quote not found for symbol: UNI7-USD"}}}
[*********************100%***********************]  29 of 29 completed
ERROR:yfinance:
4 Failed downloads:
ERROR:yfinance:['UNI7-USD']: YFPricesMissingError('possibly delisted; no price data found  (period=1y) (Yahoo error = "No data found, symbol may be delisted")')
ERROR:yfinance:['TON1-USD', 'MNT-USD', 'TAO-USD']: YFPricesMissingError('possibly delisted; no price data found  (period=1y)')


In [ ]:
# 1. Fetch 5 years of data to find the maximum possible overlap
print("Fetching 5 years of historical data...")
raw_data = yf.download(tickers, period="5y", interval="1d")['Close']

# 2. Drop any coins that are completely empty (if any failed to download)
raw_data = raw_data.dropna(axis=1, how='all')

# 3. UNIFORMING: Drop all rows (dates) where AT LEAST one coin has a missing value
# This ensures every row in your dataframe has a price for every single coin.
clean_data = raw_data.dropna()

/tmp/ipython-input-1591027290.py:3: FutureWarning: YF.download() has changed argument auto_adjust default to True
  raw_data = yf.download(tickers, period="5y", interval="1d")['Close']


Fetching 5 years of historical data...


[*******               14%                       ]  4 of 29 completedERROR:yfinance:HTTP Error 404: {"quoteSummary":{"result":null,"error":{"code":"Not Found","description":"Quote not found for symbol: UNI7-USD"}}}
[*********************100%***********************]  29 of 29 completed
ERROR:yfinance:
1 Failed download:
ERROR:yfinance:['UNI7-USD']: YFPricesMissingError('possibly delisted; no price data found  (period=5y) (Yahoo error = "No data found, symbol may be delisted")')


In [ ]:
import yfinance as yf
import pandas as pd
import numpy as np

# 1. Final Refined Ticker List (Removing UNI7-USD as requested)
tickers = [
    'BTC-USD', 'ETH-USD', 'BNB-USD', 'XRP-USD', 'SOL-USD', 'TRX-USD',
    'DOGE-USD', 'ADA-USD', 'BCH-USD', 'XMR-USD', 'LINK-USD', 'XLM-USD',
    'LTC-USD', 'AVAX-USD', 'HBAR-USD', 'SHIB-USD', 'CRO-USD',
    'DOT-USD', 'UNI-USD', 'AAVE-USD', 'OKB-USD', 'NEAR-USD', 'ETC-USD'
]

# 2. Fetch data from 01/01/2021 to 31/12/2025
# Note: end date is exclusive in yfinance, so we use 2026-01-01 to include Dec 31
raw_data = yf.download(
    tickers,
    start="2021-01-01",
    end="2026-01-01",
    interval="1d",
    auto_adjust=True
)['Close']

# 3. MBA Data Cleaning: 5-Year Date Range Validation
# We want to ensure the coins existed for the majority of this specific window
# Total expected days is roughly 1826 (365*5 + 1 leap day)
min_days = 1826 * 0.90
valid_tickers = raw_data.columns[raw_data.count() >= min_days]
five_year_data = raw_data[valid_tickers].dropna()

print(f"--- Processed Dataset (2021-2025) ---")
print(f"Final Coin Count: {len(five_year_data.columns)}")
print(f"Time Range: {five_year_data.index.min().date()} to {five_year_data.index.max().date()}")

[*********************100%***********************]  23 of 23 completed


--- Processed Dataset (2021-2025) ---
Final Coin Count: 22
Time Range: 2021-01-02 to 2025-12-31


In [ ]:
import yfinance as yf
import pandas as pd
import numpy as np

# 1. Final Refined Ticker List (Removing UNI7-USD as requested)
tickers = [
    'BTC-USD', 'ETH-USD', 'BNB-USD', 'XRP-USD', 'SOL-USD', 'TRX-USD',
    'DOGE-USD', 'ADA-USD', 'BCH-USD', 'XMR-USD', 'LINK-USD', 'XLM-USD',
    'LTC-USD', 'AVAX-USD', 'HBAR-USD', 'SHIB-USD', 'CRO-USD',
    'DOT-USD', 'UNI-USD', 'AAVE-USD', 'OKB-USD', 'NEAR-USD', 'ETC-USD'
]

# 2. Fetch data from 01/01/2021 to 31/12/2025
# Note: end date is exclusive in yfinance, so we use 2026-01-01 to include Dec 31
raw_data = yf.download(
    tickers,
    start="2021-01-01",
    end="2026-01-01",
    interval="1d",
    auto_adjust=True
)['Close']

# 3. MBA Data Cleaning: 5-Year Date Range Validation
# We want to ensure the coins existed for the majority of this specific window
# Total expected days is roughly 1826 (365*5 + 1 leap day)
min_days = 1826 * 0.90
valid_tickers = raw_data.columns[raw_data.count() >= min_days]
five_year_data = raw_data[valid_tickers].dropna()

print(f"--- Processed Dataset (2021-2025) ---")
print(f"Final Coin Count: {len(five_year_data.columns)}")
print(f"Time Range: {five_year_data.index.min().date()} to {five_year_data.index.max().date()}")

# 4. Calculate Log Prices and Log Returns
# Adding +1e-9 (epsilon) to avoid log(0) errors with SHIB
log_price = np.log(five_year_data + 1e-9)
log_return = log_price.diff().dropna()

# 5. Remove any extreme outliers (Winsorization)
# Replacing any residual 'inf' with NaN and dropping them
log_return.replace([np.inf, -np.inf], np.nan, inplace=True)
log_return = log_return.dropna()

print(f"--- Processed 5-Year Dataset ---")
print(f"Final Coin Count: {len(log_return.columns)}")
print(f"Time Range: {log_return.index.min().date()} to {log_return.index.max().date()}")

[*********************100%***********************]  23 of 23 completed


--- Processed Dataset (2021-2025) ---
Final Coin Count: 22
Time Range: 2021-01-02 to 2025-12-31
--- Processed 5-Year Dataset ---
Final Coin Count: 22
Time Range: 2021-01-03 to 2025-12-31


### Log Prices (First 5 Rows)

In [ ]:
display(log_price.head())

Ticker,AAVE-USD,ADA-USD,AVAX-USD,BCH-USD,BNB-USD,BTC-USD,CRO-USD,DOGE-USD,DOT-USD,ETC-USD,...,LINK-USD,LTC-USD,NEAR-USD,OKB-USD,SHIB-USD,SOL-USD,TRX-USD,XLM-USD,XMR-USD,XRP-USD
Date,,,,,,,,,,,,,,,,,,,,,
2021-01-02,4.451003,-1.729219,1.251316,5.870990,3.643924,10.377460,-2.783561,-4.545487,2.220164,1.754830,...,2.503085,4.919579,0.343367,2.120507,-20.723266,0.587384,-3.603840,-2.057132,4.963664,-1.506633
2021-01-03,4.511921,-1.584770,1.245003,6.044507,3.717199,10.397636,-2.817369,-4.628336,2.305908,1.880595,...,2.613752,5.076364,0.387110,2.021782,-20.723266,0.770919,-3.523534,-1.991995,4.927720,-1.487893
2021-01-05,4.757495,-1.353579,1.443953,6.039702,3.731331,10.433893,-2.712842,-4.613202,2.272297,1.980908,...,2.676894,5.066352,0.406776,2.016648,-20.723266,0.768819,-3.553230,-1.635664,4.918703,-1.483907
2021-01-09,4.767858,-1.099694,1.976651,6.352475,3.782662,10.602978,-2.490663,-4.585956,2.280111,2.050859,...,2.868326,5.178880,0.568066,1.981823,-20.723266,1.219467,-3.380630,-1.164960,4.994270,-1.118052
2021-01-10,4.837873,-1.187788,1.885021,6.402906,3.748291,10.554678,-2.553010,-4.619269,2.230557,2.185479,...,2.787935,5.142335,0.551678,1.945981,-20.723266,1.243033,-3.413219,-1.242197,5.221101,-1.144648


### Log Returns (First 5 Rows)

In [ ]:
display(log_return.head())

Ticker,AAVE-USD,ADA-USD,AVAX-USD,BCH-USD,BNB-USD,BTC-USD,CRO-USD,DOGE-USD,DOT-USD,ETC-USD,...,LINK-USD,LTC-USD,NEAR-USD,OKB-USD,SHIB-USD,SOL-USD,TRX-USD,XLM-USD,XMR-USD,XRP-USD
Date,,,,,,,,,,,,,,,,,,,,,
2021-01-03,0.060918,0.144449,-0.006314,0.173517,0.073275,0.020175,-0.033808,-0.082849,0.085744,0.125765,...,0.110667,0.156786,0.043743,-0.098725,0.0,0.183535,0.080306,0.065137,-0.035944,0.018740
2021-01-05,0.245574,0.231190,0.198950,-0.004805,0.014131,0.036258,0.104527,0.015134,-0.033611,0.100313,...,0.063142,-0.010012,0.019666,-0.005134,0.0,-0.002100,-0.029696,0.356331,-0.009017,0.003986
2021-01-09,0.010363,0.253885,0.532699,0.312773,0.051332,0.169085,0.222178,0.027246,0.007814,0.069951,...,0.191431,0.112528,0.161290,-0.034825,0.0,0.450648,0.172600,0.470703,0.075567,0.365855
2021-01-10,0.070015,-0.088094,-0.091630,0.050431,-0.034371,-0.048300,-0.062347,-0.033313,-0.049554,0.134619,...,-0.080391,-0.036545,-0.016388,-0.035842,0.0,0.023566,-0.032589,-0.077237,0.226832,-0.026596
2021-01-11,-0.087720,-0.101503,-0.014129,-0.229979,-0.100980,-0.075514,-0.151561,-0.109765,-0.117144,-0.204710,...,-0.109867,-0.206048,-0.157322,-0.098306,0.0,-0.109025,-0.128517,-0.058881,-0.154836,-0.097010


In [ ]:
correlation_matrix = log_return.corr()
print("Correlation Matrix calculated successfully.")

Correlation Matrix calculated successfully.


In [ ]:
correlation_pairs = []

for i in range(len(correlation_matrix.columns)):
    for j in range(i + 1, len(correlation_matrix.columns)): # Start from i + 1 to avoid self-correlation and duplicate pairs
        crypto1 = correlation_matrix.columns[i]
        crypto2 = correlation_matrix.columns[j]
        correlation_value = correlation_matrix.iloc[i, j]

        if correlation_value > 0: # Only consider positive correlations
            # Sort names alphabetically to ensure unique pairs regardless of order
            if crypto1 < crypto2:
                correlation_pairs.append((crypto1, crypto2, correlation_value))
            else:
                correlation_pairs.append((crypto2, crypto1, correlation_value))

# Sort the list of tuples by correlation value in descending order
sorted_correlations = sorted(correlation_pairs, key=lambda x: x[2], reverse=True)

print(f"Extracted {len(sorted_correlations)} unique positive correlation pairs.")
print("Top 5 correlated pairs:")
for pair in sorted_correlations[:5]:
    print(f"  {pair[0]} - {pair[1]}: {pair[2]:.4f}")

Extracted 231 unique positive correlation pairs.
Top 5 correlated pairs:
  BTC-USD - ETH-USD: 0.8174
  XLM-USD - XRP-USD: 0.7879
  ETH-USD - LINK-USD: 0.7775
  DOT-USD - LINK-USD: 0.7661
  AAVE-USD - ETH-USD: 0.7649


In [ ]:
non_repeating_pairs = []
selected_cryptos = set()

for crypto1, crypto2, correlation in sorted_correlations:
    if crypto1 not in selected_cryptos and crypto2 not in selected_cryptos:
        non_repeating_pairs.append((crypto1, crypto2, correlation))
        selected_cryptos.add(crypto1)
        selected_cryptos.add(crypto2)

print(f"Found {len(non_repeating_pairs)} non-repeating cryptocurrency pairs.")
print("Non-repeating pairs (sorted by correlation, highest to lowest):")
for pair in non_repeating_pairs:
    print(f"  {pair[0]} - {pair[1]}: {pair[2]:.4f}")

Found 11 non-repeating cryptocurrency pairs.
Non-repeating pairs (sorted by correlation, highest to lowest):
  BTC-USD - ETH-USD: 0.8174
  XLM-USD - XRP-USD: 0.7879
  DOT-USD - LINK-USD: 0.7661
  BCH-USD - LTC-USD: 0.7521
  ADA-USD - AVAX-USD: 0.7076
  AAVE-USD - ETC-USD: 0.6367
  BNB-USD - OKB-USD: 0.6007
  NEAR-USD - SOL-USD: 0.5995
  TRX-USD - XMR-USD: 0.5212
  CRO-USD - HBAR-USD: 0.5087
  DOGE-USD - SHIB-USD: 0.3466


In [ ]:
unique_coins_set = set()

# Select the first 5 non-repeating pairs
for crypto1, crypto2, _ in non_repeating_pairs[:5]:
    unique_coins_set.add(crypto1)
    unique_coins_set.add(crypto2)

# Convert the set to a list
unique_coins = list(unique_coins_set)

print(f"Extracted {len(unique_coins)} unique cryptocurrencies from the top 5 non-repeating pairs:")
print(unique_coins)

Extracted 10 unique cryptocurrencies from the top 5 non-repeating pairs:
['BTC-USD', 'LINK-USD', 'LTC-USD', 'ADA-USD', 'XLM-USD', 'XRP-USD', 'BCH-USD', 'DOT-USD', 'AVAX-USD', 'ETH-USD']


In [ ]:
from statsmodels.tsa.stattools import adfuller

print("Imported adfuller for ADF test.")

Imported adfuller for ADF test.


In [ ]:
adf_results = {} # Dictionary to store ADF test results

# Filter log_price for the unique coins
log_price_selected = log_price[unique_coins]

for coin in unique_coins:
    # Extract the log price series for the current coin
    series = log_price_selected[coin].dropna() # Ensure no NaNs in the series for ADF test

    if len(series) > 0: # Only perform test if series is not empty
        # Perform ADF test
        adf_test = adfuller(series, autolag='AIC')

        # Store relevant results
        adf_results[coin] = {
            'ADF Statistic': adf_test[0],
            'p-value': adf_test[1],
            'Critical Values': adf_test[4] # This is a dictionary of critical values
        }
    else:
        adf_results[coin] = {
            'ADF Statistic': np.nan,
            'p-value': np.nan,
            'Critical Values': np.nan
        }

# Convert results to a DataFrame for better readability
adf_df = pd.DataFrame.from_dict(adf_results, orient='index')

# Add a 'Stationarity' column based on p-value <= 0.05
adf_df['Stationarity'] = adf_df['p-value'].apply(lambda p: 'Stationary' if p <= 0.05 else 'Non-Stationary')

print("ADF Test Results for Log Prices of Selected Cryptocurrencies:")
display(adf_df)

ADF Test Results for Log Prices of Selected Cryptocurrencies:


,ADF Statistic,p-value,Critical Values,Stationarity
BTC-USD,-0.970258,0.763957,"{'1%': -3.4340084929858548, '5%': -2.863156021...",Non-Stationary
LINK-USD,-1.906642,0.328943,"{'1%': -3.433968010099092, '5%': -2.8631381499...",Non-Stationary
LTC-USD,-2.503235,0.114689,"{'1%': -3.4339580289794225, '5%': -2.863133743...",Non-Stationary
ADA-USD,-1.832596,0.364403,"{'1%': -3.433978046594855, '5%': -2.8631425806...",Non-Stationary
XLM-USD,-2.210934,0.202301,"{'1%': -3.433976034843506, '5%': -2.8631416925...",Non-Stationary
XRP-USD,-1.918238,0.323531,"{'1%': -3.4339881389288407, '5%': -2.863147035...",Non-Stationary
BCH-USD,-1.513901,0.526624,"{'1%': -3.4340084929858548, '5%': -2.863156021...",Non-Stationary
DOT-USD,-0.777944,0.825515,"{'1%': -3.433976034843506, '5%': -2.8631416925...",Non-Stationary
AVAX-USD,-2.610857,0.090774,"{'1%': -3.4339640110318004, '5%': -2.863136384...",Non-Stationary
ETH-USD,-2.912436,0.043921,"{'1%': -3.4339600207952796, '5%': -2.863134623...",Stationary


In [ ]:
requested_coins = ['BTC-USD', 'XLM-USD', 'XRP-USD', 'DOT-USD', 'LINK-USD', 'BCH-USD', 'LTC-USD', 'ADA-USD', 'AVAX-USD', 'ETH-USD']

# Filter log_return for the requested coins
log_return_filtered = log_return[requested_coins]

print(f"Filtered log_return_filtered DataFrame with {log_return_filtered.shape[1]} cryptocurrencies and {log_return_filtered.shape[0]} data points.")
print("Columns in log_return_filtered:", log_return_filtered.columns.tolist())
log_return_filtered.head()

Filtered log_return_filtered DataFrame with 10 cryptocurrencies and 1815 data points.
Columns in log_return_filtered: ['BTC-USD', 'XLM-USD', 'XRP-USD', 'DOT-USD', 'LINK-USD', 'BCH-USD', 'LTC-USD', 'ADA-USD', 'AVAX-USD', 'ETH-USD']


Ticker,BTC-USD,XLM-USD,XRP-USD,DOT-USD,LINK-USD,BCH-USD,LTC-USD,ADA-USD,AVAX-USD,ETH-USD
Date,,,,,,,,,,
2021-01-03,0.020175,0.065137,0.018740,0.085744,0.110667,0.173517,0.156786,0.144449,-0.006314,0.230695
2021-01-05,0.036258,0.356331,0.003986,-0.033611,0.063142,-0.004805,-0.010012,0.231190,0.198950,0.120113
2021-01-09,0.169085,0.470703,0.365855,0.007814,0.191431,0.312773,0.112528,0.253885,0.532699,0.152386
2021-01-10,-0.048300,-0.077237,-0.026596,-0.049554,-0.080391,0.050431,-0.036545,-0.088094,-0.091630,-0.014808
2021-01-11,-0.075514,-0.058881,-0.097010,-0.117144,-0.109867,-0.229979,-0.206048,-0.101503,-0.014129,-0.146582


In [ ]:
adf_results_returns = {} # Initialize an empty dictionary to store ADF test results for log returns

for coin in log_return_filtered.columns:
    # Extract the log return series for the current coin
    series = log_return_filtered[coin].dropna() # Ensure no NaNs in the series for ADF test

    if len(series) > 0: # Only perform test if series is not empty
        # Perform ADF test
        adf_test = adfuller(series, autolag='AIC')

        # Store relevant results
        adf_results_returns[coin] = {
            'ADF Statistic': adf_test[0],
            'p-value': adf_test[1],
            'Critical Values': adf_test[4] # This is a dictionary of critical values
        }
    else:
        adf_results_returns[coin] = {
            'ADF Statistic': np.nan,
            'p-value': np.nan,
            'Critical Values': np.nan
        }

print("ADF Test results for log returns stored in 'adf_results_returns'.")

ADF Test results for log returns stored in 'adf_results_returns'.


In [ ]:
adf_df_returns_filtered = pd.DataFrame.from_dict(adf_results_returns, orient='index')

# Add a 'Stationarity' column based on p-value <= 0.05
adf_df_returns_filtered['Stationarity'] = adf_df_returns_filtered['p-value'].apply(lambda p: 'Stationary' if p <= 0.05 else 'Non-Stationary')

print("ADF Test Results for Log Returns of Selected Cryptocurrencies:")
display(adf_df_returns_filtered)

ADF Test Results for Log Returns of Selected Cryptocurrencies:


,ADF Statistic,p-value,Critical Values,Stationarity
BTC-USD,-7.986304,2.543552e-12,"{'1%': -3.4340084929858548, '5%': -2.863156021...",Stationary
XLM-USD,-12.140289,1.657015e-22,"{'1%': -3.433976034843506, '5%': -2.8631416925...",Stationary
XRP-USD,-9.994207,1.959729e-17,"{'1%': -3.4339881389288407, '5%': -2.863147035...",Stationary
DOT-USD,-14.648568,3.530253e-27,"{'1%': -3.433976034843506, '5%': -2.8631416925...",Stationary
LINK-USD,-20.328152,0.000000e+00,"{'1%': -3.433968010099092, '5%': -2.8631381499...",Stationary
BCH-USD,-8.433986,1.837722e-13,"{'1%': -3.4340084929858548, '5%': -2.863156021...",Stationary
LTC-USD,-12.868985,4.925289e-24,"{'1%': -3.4339861159726306, '5%': -2.863146142...",Stationary
ADA-USD,-12.549681,2.220134e-23,"{'1%': -3.433978046594855, '5%': -2.8631425806...",Stationary
AVAX-USD,-23.207238,0.000000e+00,"{'1%': -3.4339640110318004, '5%': -2.863136384...",Stationary
ETH-USD,-44.423725,0.000000e+00,"{'1%': -3.4339600207952796, '5%': -2.863134623...",Stationary


In [ ]:
selected_non_stationary_coins = adf_df[adf_df['p-value'] > 0.05].index.tolist()

print(f"Cryptocurrencies with non-stationary log prices (p-value > 0.05): {selected_non_stationary_coins}")

Cryptocurrencies with non-stationary log prices (p-value > 0.05): ['BTC-USD', 'LINK-USD', 'LTC-USD', 'ADA-USD', 'XLM-USD', 'XRP-USD', 'BCH-USD', 'DOT-USD', 'AVAX-USD']


In [ ]:
cointegration_pairs = []

for crypto1, crypto2, _ in non_repeating_pairs:
    if crypto1 in selected_non_stationary_coins and crypto2 in selected_non_stationary_coins:
        cointegration_pairs.append((crypto1, crypto2))

print(f"Selected {len(cointegration_pairs)} pairs for cointegration testing:")
for pair in cointegration_pairs:
    print(f"  {pair[0]} - {pair[1]}")

Selected 4 pairs for cointegration testing:
  XLM-USD - XRP-USD
  DOT-USD - LINK-USD
  BCH-USD - LTC-USD
  ADA-USD - AVAX-USD


# Task
Confirm that the log prices for the cryptocurrencies in `cointegration_pairs` are I(1) by verifying that their log returns (first differences) are stationary, leveraging the results from the `adf_df_returns_filtered` DataFrame.

## Confirm I(1) Property of Log Prices

### Subtask:
For the cryptocurrencies in the `cointegration_pairs` list, verify that their log prices are integrated of order 1 (I(1)). This is done by confirming that their first differences (log returns) are stationary. We will leverage the results from the `adf_df_returns_filtered` DataFrame generated earlier, which contains ADF test results for log returns, to confirm this prerequisite for the Engle-Granger test.


**Reasoning**:
First, I'll extract all unique cryptocurrency tickers from the `cointegration_pairs` list to identify the specific coins relevant for this check. This is necessary to filter the ADF test results for log returns.



In [ ]:
unique_coins_for_cointegration = set()
for crypto1, crypto2 in cointegration_pairs:
    unique_coins_for_cointegration.add(crypto1)
    unique_coins_for_cointegration.add(crypto2)

# Convert set to a list
unique_coins_for_cointegration_list = list(unique_coins_for_cointegration)

print(f"Unique cryptocurrencies identified from cointegration pairs: {unique_coins_for_cointegration_list}")

Unique cryptocurrencies identified from cointegration pairs: ['LTC-USD', 'LINK-USD', 'ADA-USD', 'XLM-USD', 'XRP-USD', 'BCH-USD', 'DOT-USD', 'AVAX-USD']


**Reasoning**:
Now that the unique cryptocurrencies from the cointegration pairs have been identified, I will filter the `adf_df_returns_filtered` DataFrame to show only the ADF test results for their log returns. This will allow us to confirm if their log returns are stationary, which implies their log prices are I(1).



In [ ]:
filtered_returns_stationarity = adf_df_returns_filtered.loc[unique_coins_for_cointegration_list]

print("ADF Test Results for Log Returns of Cryptocurrencies in Cointegration Pairs:")
display(filtered_returns_stationarity)

# Confirm all are stationary
all_stationary = filtered_returns_stationarity['Stationarity'].eq('Stationary').all()

if all_stationary:
    print("\nAll cryptocurrencies in cointegration pairs have stationary log returns (I(1)).")
else:
    print("\nWARNING: Not all cryptocurrencies in cointegration pairs have stationary log returns.")

ADF Test Results for Log Returns of Cryptocurrencies in Cointegration Pairs:


,ADF Statistic,p-value,Critical Values,Stationarity
LTC-USD,-12.868985,4.925289e-24,"{'1%': -3.4339861159726306, '5%': -2.863146142...",Stationary
LINK-USD,-20.328152,0.000000e+00,"{'1%': -3.433968010099092, '5%': -2.8631381499...",Stationary
ADA-USD,-12.549681,2.220134e-23,"{'1%': -3.433978046594855, '5%': -2.8631425806...",Stationary
XLM-USD,-12.140289,1.657015e-22,"{'1%': -3.433976034843506, '5%': -2.8631416925...",Stationary
XRP-USD,-9.994207,1.959729e-17,"{'1%': -3.4339881389288407, '5%': -2.863147035...",Stationary
BCH-USD,-8.433986,1.837722e-13,"{'1%': -3.4340084929858548, '5%': -2.863156021...",Stationary
DOT-USD,-14.648568,3.530253e-27,"{'1%': -3.433976034843506, '5%': -2.8631416925...",Stationary
AVAX-USD,-23.207238,0.000000e+00,"{'1%': -3.4339640110318004, '5%': -2.863136384...",Stationary



All cryptocurrencies in cointegration pairs have stationary log returns (I(1)).


In [ ]:
from statsmodels.formula.api import ols
from statsmodels.tsa.stattools import adfuller

# Initialize a dictionary to store Engle-Granger test results
engle_granger_results = {}

# Filter log_price to include only the relevant coins for cointegration testing
# This ensures that we are only working with the log prices of the coins identified as non-stationary in their log prices
# and part of the cointegration_pairs.

# First, gather all unique coins from cointegration_pairs
all_coins_in_cointegration_pairs = list(unique_coins_for_cointegration)
log_price_for_cointegration = log_price[all_coins_in_cointegration_pairs]

print("Starting Engle-Granger Cointegration Test for selected pairs...")

for crypto1, crypto2 in cointegration_pairs:
    # 1. Align their log price series for the current pair
    # Create a temporary DataFrame for the current pair, ensuring no NaNs
    df_pair = log_price_for_cointegration[[crypto1, crypto2]].dropna()

    if df_pair.empty:
        print(f"Warning: No overlapping data for {crypto1} and {crypto2}. Skipping.")
        engle_granger_results[(crypto1, crypto2)] = {'ADF Statistic': np.nan, 'p-value': np.nan, 'Critical Values': np.nan, 'Cointegrated': False}
        continue

    # Rename columns to be compatible with patsy/ols formula (replace hyphens with underscores)
    df_pair_renamed = df_pair.copy()
    renamed_crypto1 = crypto1.replace('-', '_')
    renamed_crypto2 = crypto2.replace('-', '_')
    df_pair_renamed.rename(columns={crypto1: renamed_crypto1, crypto2: renamed_crypto2}, inplace=True)

    # 2. Perform OLS regression
    # Regress crypto1's log price on crypto2's log price using renamed columns
    model = ols(f'{renamed_crypto1} ~ {renamed_crypto2}', data=df_pair_renamed).fit()

    # 3. Extract the residuals from the OLS regression
    residuals = model.resid

    # 4. Apply the Augmented Dickey-Fuller (ADF) test to these residuals
    adf_test_residuals = adfuller(residuals, autolag='AIC')

    # 5. Store the results
    p_value_residuals = adf_test_residuals[1]
    is_cointegrated = p_value_residuals <= 0.05 # Check for stationarity of residuals

    engle_granger_results[(crypto1, crypto2)] = {
        'ADF Statistic': adf_test_residuals[0],
        'p-value': p_value_residuals,
        'Critical Values': adf_test_residuals[4],
        'Cointegrated': is_cointegrated
    }

print("Engle-Granger Cointegration Test complete. Results stored in 'engle_granger_results'.")

Starting Engle-Granger Cointegration Test for selected pairs...
Engle-Granger Cointegration Test complete. Results stored in 'engle_granger_results'.


In [ ]:
engle_granger_df = pd.DataFrame.from_dict(engle_granger_results, orient='index')
engle_granger_df.index.names = ['Crypto1', 'Crypto2']

print("Engle-Granger Cointegration Test Results:")
display(engle_granger_df)

# Filter to show only cointegrated pairs
cointegrated_pairs_df = engle_granger_df[engle_granger_df['Cointegrated'] == True]

print("\nCointegrated Pairs (based on p-value <= 0.05 for residuals):")
display(cointegrated_pairs_df)

Engle-Granger Cointegration Test Results:


,,ADF Statistic,p-value,Critical Values,Cointegrated
Crypto1,Crypto2,,,,
XLM-USD,XRP-USD,-2.834499,0.053512,"{'1%': -3.433978046594855, '5%': -2.8631425806...",False
DOT-USD,LINK-USD,-0.084481,0.950963,"{'1%': -3.4339580289794225, '5%': -2.863133743...",False
BCH-USD,LTC-USD,-1.013147,0.748439,"{'1%': -3.4339600207952796, '5%': -2.863134623...",False
ADA-USD,AVAX-USD,-1.658960,0.452372,"{'1%': -3.4339580289794225, '5%': -2.863133743...",False



Cointegrated Pairs (based on p-value <= 0.05 for residuals):


,,ADF Statistic,p-value,Critical Values,Cointegrated
Crypto1,Crypto2,,,,


In [ ]:
import statsmodels.api as sm

# Define the four pairs for OLS regression
ols_pairs_log_returns = [
    ('XLM-USD', 'XRP-USD'),
    ('DOT-USD', 'LINK-USD'),
    ('BCH-USD', 'LTC-USD'),
    ('ADA-USD', 'AVAX-USD')
]

print("Performing OLS regression for each specified pair using log returns:")

for dep_crypto, indep_crypto in ols_pairs_log_returns:
    # Ensure the data is aligned and drop any NaNs for the current pair
    df_pair_returns = log_return[[dep_crypto, indep_crypto]].dropna()

    if df_pair_returns.empty:
        print(f"Warning: No overlapping log return data for {dep_crypto} and {indep_crypto}. Skipping.")
        continue

    # Define dependent (Y) and independent (X) variables
    Y = df_pair_returns[dep_crypto]
    X = df_pair_returns[indep_crypto]
    X = sm.add_constant(X) # Add a constant term for the intercept

    # Run the OLS regression
    model = sm.OLS(Y, X)
    results = model.fit()

    # Print the summary of the regression results for the current pair
    print(f"\n--- OLS Regression Results for {dep_crypto} (Dependent) vs {indep_crypto} (Independent) ---")
    print(results.summary())


Performing OLS regression for each specified pair using log returns:

--- OLS Regression Results for XLM-USD (Dependent) vs XRP-USD (Independent) ---
                            OLS Regression Results                            
Dep. Variable:                XLM-USD   R-squared:                       0.621
Model:                            OLS   Adj. R-squared:                  0.621
Method:                 Least Squares   F-statistic:                     2967.
Date:                Thu, 12 Feb 2026   Prob (F-statistic):               0.00
Time:                        01:30:49   Log-Likelihood:                 3754.6
No. Observations:                1815   AIC:                            -7505.
Df Residuals:                    1813   BIC:                            -7494.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      

In [ ]:
swapped_ols_pairs_log_returns = [
    ('XRP-USD', 'XLM-USD'),
    ('LINK-USD', 'DOT-USD'),
    ('LTC-USD', 'BCH-USD'),
    ('AVAX-USD', 'ADA-USD')
]

print("Defined swapped_ols_pairs_log_returns:")
print(swapped_ols_pairs_log_returns)

Defined swapped_ols_pairs_log_returns:
[('XRP-USD', 'XLM-USD'), ('LINK-USD', 'DOT-USD'), ('LTC-USD', 'BCH-USD'), ('AVAX-USD', 'ADA-USD')]


In [ ]:
import statsmodels.api as sm

ols_summary_results_swapped = {}

print("Performing OLS regression for each specified swapped pair using log returns:")

for dep_crypto, indep_crypto in swapped_ols_pairs_log_returns:
    # Ensure the data is aligned and drop any NaNs for the current pair
    df_pair_returns = log_return[[dep_crypto, indep_crypto]].dropna()

    if df_pair_returns.empty:
        print(f"Warning: No overlapping log return data for {dep_crypto} and {indep_crypto}. Skipping.")
        ols_summary_results_swapped[f'{dep_crypto} vs {indep_crypto}'] = {
            'Independent Variable Coefficient': np.nan,
            'Independent Variable P-value': np.nan,
            'Intercept Coefficient': np.nan,
            'Intercept P-value': np.nan,
            'R-squared': np.nan
        }
        continue

    # Define dependent (Y) and independent (X) variables
    Y = df_pair_returns[dep_crypto]
    X = df_pair_returns[indep_crypto]
    X = sm.add_constant(X) # Add a constant term for the intercept

    # Run the OLS regression
    model = sm.OLS(Y, X)
    results = model.fit()

    # Store key summary statistics
    ols_summary_results_swapped[f'{dep_crypto} vs {indep_crypto}'] = {
        'Independent Variable Coefficient': results.params[indep_crypto],
        'Independent Variable P-value': results.pvalues[indep_crypto],
        'Intercept Coefficient': results.params['const'],
        'Intercept P-value': results.pvalues['const'],
        'R-squared': results.rsquared
    }

    # Print a concise summary
    print(f"\n--- OLS Regression Results for {dep_crypto} (Dependent) vs {indep_crypto} (Independent) ---")
    print(f"  Independent Variable Coefficient: {results.params[indep_crypto]:.4f} (P-value: {results.pvalues[indep_crypto]:.4f})")
    print(f"  Intercept Coefficient: {results.params['const']:.4f} (P-value: {results.pvalues['const']:.4f})")
    print(f"  R-squared: {results.rsquared:.4f}")

print("\nOLS regression for swapped pairs complete. Results stored in 'ols_summary_results_swapped'.")

Performing OLS regression for each specified swapped pair using log returns:

--- OLS Regression Results for XRP-USD (Dependent) vs XLM-USD (Independent) ---
  Independent Variable Coefficient: 0.8241 (P-value: 0.0000)
  Intercept Coefficient: 0.0010 (P-value: 0.2008)
  R-squared: 0.6207

--- OLS Regression Results for LINK-USD (Dependent) vs DOT-USD (Independent) ---
  Independent Variable Coefficient: 0.7768 (P-value: 0.0000)
  Intercept Coefficient: 0.0007 (P-value: 0.3821)
  R-squared: 0.5869

--- OLS Regression Results for LTC-USD (Dependent) vs BCH-USD (Independent) ---
  Independent Variable Coefficient: 0.6938 (P-value: 0.0000)
  Intercept Coefficient: -0.0005 (P-value: 0.4626)
  R-squared: 0.5656

--- OLS Regression Results for AVAX-USD (Dependent) vs ADA-USD (Independent) ---
  Independent Variable Coefficient: 0.8724 (P-value: 0.0000)
  Intercept Coefficient: 0.0004 (P-value: 0.7048)
  R-squared: 0.5008

OLS regression for swapped pairs complete. Results stored in 'ols_summa

In [ ]:
granger_causality_pairs = [
    ('XRP-USD', 'XLM-USD'),
    ('LINK-USD', 'DOT-USD'),
    ('LTC-USD', 'BCH-USD'),
    ('AVAX-USD', 'ADA-USD')
]

print("Defined granger_causality_pairs:")
print(granger_causality_pairs)

Defined granger_causality_pairs:
[('XRP-USD', 'XLM-USD'), ('LINK-USD', 'DOT-USD'), ('LTC-USD', 'BCH-USD'), ('AVAX-USD', 'ADA-USD')]


In [ ]:
from statsmodels.tsa.api import VAR
from statsmodels.tsa.stattools import grangercausalitytests

print("Imported VAR and grangercausalitytests for Granger causality testing.")

Imported VAR and grangercausalitytests for Granger causality testing.


In [ ]:
granger_results = {}

# Max lags to consider for the Granger Causality test
# A common choice is 1 to 5, given our daily data, 5 lags seems reasonable.
# We will use 5 lags as a starting point. Adjust if necessary based on data characteristics.
max_lags = 5

print(f"Performing Granger Causality tests for {len(granger_causality_pairs)} pairs with max_lags={max_lags}...")

for crypto_y, crypto_x in granger_causality_pairs:
    # Prepare the data for Granger causality test
    # The function expects a DataFrame with the dependent variable as the first column
    # and the independent variable as the second column.
    data_for_granger = log_return[[crypto_y, crypto_x]].dropna()

    if data_for_granger.empty:
        print(f"Warning: No overlapping data for {crypto_y} and {crypto_x}. Skipping pair.")
        continue

    # Test direction 1: Does X Granger-cause Y?
    # H0: X does not Granger-cause Y
    # The function returns a dictionary of test results for each lag
    test_result_x_to_y = grangercausalitytests(data_for_granger[[crypto_y, crypto_x]], max_lags, verbose=False)
    # Get the p-value for the F-test at the max_lags
    p_value_x_to_y = test_result_x_to_y[max_lags][0]['ssr_ftest'][1] # F-test p-value

    # Test direction 2: Does Y Granger-cause X?
    # H0: Y does not Granger-cause X
    data_for_granger_swapped = log_return[[crypto_x, crypto_y]].dropna()
    test_result_y_to_x = grangercausalitytests(data_for_granger_swapped[[crypto_x, crypto_y]], max_lags, verbose=False)
    p_value_y_to_x = test_result_y_to_x[max_lags][0]['ssr_ftest'][1] # F-test p-value

    granger_results[f'{crypto_x} -> {crypto_y}'] = p_value_x_to_y
    granger_results[f'{crypto_y} -> {crypto_x}'] = p_value_y_to_x

print("Granger Causality tests complete. Results stored in 'granger_results'.")

# Display the results for quick inspection
for test, p_value in granger_results.items():
    print(f"{test}: p-value = {p_value:.4f}")

Performing Granger Causality tests for 4 pairs with max_lags=5...


/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print resul

Granger Causality tests complete. Results stored in 'granger_results'.
XLM-USD -> XRP-USD: p-value = 0.0814
XRP-USD -> XLM-USD: p-value = 0.0002
DOT-USD -> LINK-USD: p-value = 0.0072
LINK-USD -> DOT-USD: p-value = 0.3061
BCH-USD -> LTC-USD: p-value = 0.5345
LTC-USD -> BCH-USD: p-value = 0.2254
ADA-USD -> AVAX-USD: p-value = 0.1880
AVAX-USD -> ADA-USD: p-value = 0.1784


/usr/local/lib/python3.12/dist-packages/statsmodels/tsa/stattools.py:1556: FutureWarning: verbose is deprecated since functions should not print results
  warnings.warn(


In [ ]:
granger_results = {}

# Max lags to consider for the Granger Causality test
# A common choice is 1 to 5, given our daily data, 5 lags seems reasonable.
# We will use 5 lags as a starting point. Adjust if necessary based on data characteristics.
max_lags = 5

print(f"Performing Granger Causality tests for {len(granger_causality_pairs)} pairs with max_lags={max_lags}...")

for crypto_y, crypto_x in granger_causality_pairs:
    # Prepare the data for Granger causality test
    # The function expects a DataFrame with the dependent variable as the first column
    # and the independent variable as the second column.
    data_for_granger = log_return[[crypto_y, crypto_x]].dropna()

    if data_for_granger.empty:
        print(f"Warning: No overlapping data for {crypto_y} and {crypto_x}. Skipping pair.")
        continue

    # Test direction 1: Does X Granger-cause Y?
    # H0: X does not Granger-cause Y
    # The function returns a dictionary of test results for each lag
    test_result_x_to_y = grangercausalitytests(data_for_granger[[crypto_y, crypto_x]], max_lags)
    # Get the p-value for the F-test at the max_lags
    p_value_x_to_y = test_result_x_to_y[max_lags][0]['ssr_ftest'][1] # F-test p-value

    # Test direction 2: Does Y Granger-cause X?
    # H0: Y does not Granger-cause X
    data_for_granger_swapped = log_return[[crypto_x, crypto_y]].dropna()
    test_result_y_to_x = grangercausalitytests(data_for_granger_swapped[[crypto_x, crypto_y]], max_lags)
    p_value_y_to_x = test_result_y_to_x[max_lags][0]['ssr_ftest'][1] # F-test p-value

    granger_results[f'{crypto_x} -> {crypto_y}'] = p_value_x_to_y
    granger_results[f'{crypto_y} -> {crypto_x}'] = p_value_y_to_x

print("Granger Causality tests complete. Results stored in 'granger_results'.")

# Display the results for quick inspection
for test, p_value in granger_results.items():
    print(f"{test}: p-value = {p_value:.4f}")

Performing Granger Causality tests for 4 pairs with max_lags=5...

Granger Causality
number of lags (no zero) 1
ssr based F test:         F=2.6905  , p=0.1011  , df_denom=1811, df_num=1
ssr based chi2 test:   chi2=2.6950  , p=0.1007  , df=1
likelihood ratio test: chi2=2.6930  , p=0.1008  , df=1
parameter F test:         F=2.6905  , p=0.1011  , df_denom=1811, df_num=1

Granger Causality
number of lags (no zero) 2
ssr based F test:         F=4.7863  , p=0.0084  , df_denom=1808, df_num=2
ssr based chi2 test:   chi2=9.5990  , p=0.0082  , df=2
likelihood ratio test: chi2=9.5737  , p=0.0083  , df=2
parameter F test:         F=4.7863  , p=0.0084  , df_denom=1808, df_num=2

Granger Causality
number of lags (no zero) 3
ssr based F test:         F=2.4732  , p=0.0600  , df_denom=1805, df_num=3
ssr based chi2 test:   chi2=7.4483  , p=0.0589  , df=3
likelihood ratio test: chi2=7.4331  , p=0.0593  , df=3
parameter F test:         F=2.4732  , p=0.0600  , df_denom=1805, df_num=3

Granger Causality
num

### Summarize Granger Causality Test Results

Based on the `granger_results` dictionary, we can summarize the Granger causality relationships for the log returns of the specified pairs. We will typically use a significance level of 0.05 (alpha = 5%) to determine if a null hypothesis (no Granger causality) can be rejected.

Here's a breakdown of the results:

*   **XLM-USD -> XRP-USD**: p-value = 0.0814. Since p > 0.05, we **do not reject** the null hypothesis. There is no significant evidence that XLM-USD Granger-causes XRP-USD.
*   **XRP-USD -> XLM-USD**: p-value = 0.0002. Since p < 0.05, we **reject** the null hypothesis. There is significant evidence that XRP-USD Granger-causes XLM-USD.

*   **DOT-USD -> LINK-USD**: p-value = 0.0072. Since p < 0.05, we **reject** the null hypothesis. There is significant evidence that DOT-USD Granger-causes LINK-USD.
*   **LINK-USD -> DOT-USD**: p-value = 0.3061. Since p > 0.05, we **do not reject** the null hypothesis. There is no significant evidence that LINK-USD Granger-causes DOT-USD.

*   **BCH-USD -> LTC-USD**: p-value = 0.5345. Since p > 0.05, we **do not reject** the null hypothesis. There is no significant evidence that BCH-USD Granger-causes LTC-USD.
*   **LTC-USD -> BCH-USD**: p-value = 0.2254. Since p > 0.05, we **do not reject** the null hypothesis. There is no significant evidence that LTC-USD Granger-causes BCH-USD.

*   **ADA-USD -> AVAX-USD**: p-value = 0.1880. Since p > 0.05, we **do not reject** the null hypothesis. There is no significant evidence that ADA-USD Granger-causes AVAX-USD.
*   **AVAX-USD -> ADA-USD**: p-value = 0.1784. Since p > 0.05, we **do not reject** the null hypothesis. There is no significant evidence that AVAX-USD Granger-causes ADA-USD.

**Conclusion:**

*   XRP-USD Granger-causes XLM-USD.
*   DOT-USD Granger-causes LINK-USD.
*   No significant Granger causality was found in the other tested directions or pairs at the 0.05 significance level.

In [ ]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.tsa.ardl import ardl_select_order, UECM

# 1. Prepare data
df_pair = log_price[['ETH-USD', 'BTC-USD']].dropna()
df_pair.index = pd.to_datetime(df_pair.index)
df_pair = df_pair.asfreq('D').ffill()

# 2. Select the optimal ARDL order
sel_res = ardl_select_order(
    df_pair['ETH-USD'],
    maxlag=5,
    exog=df_pair[['BTC-USD']],
    maxorder=5,
    ic="aic",
    trend="c",
    glob=True
)

# Extract the lags from the selection result
lags = sel_res.model.ardl_order # This is the (p, q) tuple, e.g., (4, 5)
print(f"The optimal ARDL order is: {lags}")
# 3. Manually initialize and fit the UECM using the optimal lags
# UECM(endog, lags, exog, order, trend)
# order corresponds to the lags of the independent variable (BTC)
uecm_model = UECM(
    df_pair['ETH-USD'],
    lags[0],
    exog=df_pair[['BTC-USD']],
    order=lags[1],
    trend="c"
)
uecm_res = uecm_model.fit()
# 3. Manually initialize and fit the UECM using the optimal lags
# UECM(endog, lags, exog, order, trend)
# order corresponds to the lags of the independent variable (BTC)
uecm_model = UECM(
    df_pair['ETH-USD'],
    lags[0],
    exog=df_pair[['BTC-USD']],
    order=lags[1],
    trend="c"
)
uecm_res = uecm_model.fit()

The optimal ARDL order is: (4, 5)


In [ ]:
# ... (Assuming the previous UECM model was fitted as uecm_res)

# 4. Perform the ARDL Bounds Test
bounds_test_result = uecm_res.bounds_test(case=3)

# 5. Display Results
print("="*55)
print(f"{'ARDL BOUNDS TEST (COINTEGRATION)':^55}")
print("="*55)
print(f"F-statistic: {bounds_test_result.stat:.4f}")
print("-" * 55)

# The crit_vals is a DataFrame, so we print it directly or iterate its rows
print("Critical Values Table:")
print(bounds_test_result.crit_vals)

print("-" * 55)

# 6. Final Interpretation Logic
f_stat = bounds_test_result.stat
# Accessing the 5% row and the I(1) column (index 1)
# Note: crit_vals index is usually float [0.1, 0.05, 0.01] or [10, 5, 1]
try:
    # Try accessing by common pandas index for statsmodels
    upper_bound_5pct = bounds_test_result.crit_vals.loc[0.05, 1]
except:
    # Fallback: take the second value of the second row
    upper_bound_5pct = bounds_test_result.crit_vals.iloc[1, 1]

if f_stat > upper_bound_5pct:
    print(f"CONCLUSION: Cointegration Confirmed ({f_stat:.2f} > {upper_bound_5pct:.2f})")
else:
    print(f"CONCLUSION: No Cointegration Detected ({f_stat:.2f} <= {upper_bound_5pct:.2f})")
    print("Interpretation: BTC and ETH do not have a stable long-run equilibrium in this period.")

# 7. Show the UECM Coefficients (To see the ECT)
print("\n" + "="*55)
print(f"{'UECM MODEL SUMMARY':^55}")
print("="*55)
print(uecm_res.summary())

           ARDL BOUNDS TEST (COINTEGRATION)            
F-statistic: 1.2235
-------------------------------------------------------
Critical Values Table:
               lower     upper
percentile                    
90.0        3.174128  4.108262
95.0        3.802263  4.811686
99.0        5.172288  6.322055
99.9        7.024122  8.322478
-------------------------------------------------------
CONCLUSION: No Cointegration Detected (1.22 <= 4.81)
Interpretation: BTC and ETH do not have a stable long-run equilibrium in this period.

                  UECM MODEL SUMMARY                   
                              UECM Model Results                              
Dep. Variable:              D.ETH-USD   No. Observations:                 1825
Model:                     UECM(4, 5)   Log Likelihood                4283.936
Method:               Conditional MLE   S.D. of innovations              7.780
Date:                Thu, 12 Feb 2026   AIC                          -8543.872
Time:       